File: src/leetCodeProblems/DFSUsingRecursion/BinaryTree/BinaryTreeDiameter543.java
package leetCodeProblems.BinaryTree;

/**
 * LeetCode - https://leetcode.com/problems/diameter-of-binary-tree/
 */

public class BinaryTreeDiameter543 {

    public int depth(TreeNode node) {

        if (node == null) {
            return 0;
        }

        //System.out.println("Current node value -> " + node.val);
        int leftChildHeight = depth(node.left);
        int rightChildHeight = depth(node.right);

        //System.out.println("LeftChildHeight => " + leftChildHeight);
        //System.out.println("RightChildHeight => " + rightChildHeight);

        int currentNodeHeight = 1 + Math.max(leftChildHeight, rightChildHeight);

        //System.out.println("Current node value -> " + node.val);
        //System.out.println("currentNodeHeight => " + currentNodeHeight);

        //System.out.println("-----");
        return currentNodeHeight;
    }

    public int diameter(TreeNode node) {

        if (node == null) {
            return 0;
        }

        //int height = depth(node) - 1;
        int leftChildHeight = depth(node.left) - 1;
        int rightChildHeight = depth(node.right) -1;

        int leftChildDiameter = diameter(node.left);
        int rightChildDiameter = diameter(node.right);

        return Math.max(leftChildHeight + rightChildHeight + 2, Math.max(leftChildDiameter, rightChildDiameter));
    }

    static class TreeNode {

        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }

    }

    public static void main(String[] args) {

        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(3);
        root.left.left = new TreeNode(4);
        root.left.right = new TreeNode(5);
        root.right.left = new TreeNode(6);
        root.right.right = new TreeNode(7);

        BinaryTreeDiameter543 ob = new BinaryTreeDiameter543();

        int diameter = ob.diameter(root);

        System.out.println(diameter);

    }
}

File: src/leetCodeProblems/DFSUsingRecursion/BinaryTree/BinaryTreeDiameter543.go
package main

/*
- Leetcode - https://leetcode.com/problems/diameter-of-binary-tree/
- Time - O(n^2)
- Space - O(1)
*/

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func heightUtil(node *TreeNode, prevHeight int, maxHeight *int) {
	if node == nil {
		return
	}

	if prevHeight > *maxHeight {
		*maxHeight = prevHeight
	}

	prevHeight++

	heightUtil(node.Left, prevHeight, maxHeight)
	heightUtil(node.Right, prevHeight, maxHeight)
}

func height(node *TreeNode) int {
	out := 0
	heightUtil(node, 0, &out)
	return out
}

func diameterOfBinaryTree(root *TreeNode) int {
	if root == nil {
		return 0
	}

	leftHeight, rightHeight := 0, 0

	if root.Left != nil {
		leftHeight = height(root.Left) + 1
	}

	if root.Right != nil {
		rightHeight = height(root.Right) + 1
	}

	return max(max(diameterOfBinaryTree(root.Left), diameterOfBinaryTree(root.Right)),
		leftHeight+rightHeight)
}

// func main() {
// 	// root := new(TreeNode) //Returns pointer to TreeNode object
// 	// root.Val = 4
// 	// root.Left = new(TreeNode)
// 	// root.Left.Val = 2
// 	// root.Left.Left = new(TreeNode)
// 	// root.Left.Left.Val = 1
// 	// root.Left.Right = new(TreeNode)
// 	// root.Left.Right.Val = 3

// 	// root.Right = new(TreeNode)
// 	// root.Right.Val = 7
// 	// root.Right.Left = new(TreeNode)
// 	// root.Right.Left.Val = 6
// 	// root.Right.Right = new(TreeNode)
// 	// root.Right.Right.Val = 9

// 	// // maxHeight := 0
// 	// // height(root, 0, &maxHeight)
// 	// // log.Println(maxHeight)

// 	// log.Println(diameterOfBinaryTree(root))

// 	root := new(TreeNode) //Returns pointer to TreeNode object
// 	root.Val = 1
// 	root.Left = new(TreeNode)
// 	root.Left.Val = 2
// 	root.Left.Left = new(TreeNode)
// 	root.Left.Left.Val = 4
// 	root.Left.Right = new(TreeNode)
// 	root.Left.Right.Val = 5

// 	root.Right = new(TreeNode)
// 	root.Right.Val = 3

// 	// maxHeight := 0
// 	// height(root, 0, &maxHeight)
// 	// log.Println(maxHeight)

// 	log.Println(diameterOfBinaryTree(root))
// }
